src/txn/txn_recover.c

```diff
@@ -822,12 +846,164 @@ __recovery_file_scan_prefix(WT_RECOVERY *r, const char *prefix, const char *igno
+ *     them for recovery.
+ */
+static int
+__recovery_file_scan_prefix(WT_RECOVERY *r, const char *prefix, const char *ignore_suffix)
+{
+    return (__recovery_metadata_iterate_func(r, prefix, ignore_suffix, __recovery_setup_file));
+}
+
+/*
+ * __metadata_check_consistency --
+ *     For each table metadata entry, check that the associated colgroup and file metadata entries
+ *     exist. If they do not exist, mark them for removal.
+ */
+static int
+__metadata_check_consistency(WT_RECOVERY *r, const char *uri, const char *config)
+{
+    WT_CONFIG cparser;
+    WT_CONFIG_ITEM ckey, cval, config_item;
+    WT_CURSOR *metac;
+    WT_DECL_RET;
+    size_t len;
+    char *cgname, *filename, *tieredname;
+    const char *format_name;
+    bool has_colgroup, has_file;
+
+    has_file = has_colgroup = true;
+    cgname = filename = tieredname = NULL;
+    format_name = uri;
+    WT_PREFIX_SKIP_REQUIRED(r->session, format_name, "table:");
+
+    WT_ERR(__wt_metadata_cursor_open(r->session, NULL, &metac));
+    /*
+     * FIXME-WT-XXXX: Add capability for cleaning complex and tiered tables. For now, only simple
+     * tables are considered.
+     */
+    WT_ERR(__wt_config_getones(r->session, config, "columns", &config_item));
+    __wt_config_subinit(r->session, &cparser, &config_item);
+    if ((ret = __wt_config_next(&cparser, &ckey, &cval)) == 0)
+        goto done;
+
+    len = strlen("tiered:") + strlen(format_name) + 1;
+    WT_ERR(__wt_calloc_def(r->session, len, &tieredname));
+    WT_ERR(__wt_snprintf(tieredname, len, "tiered:%s", format_name));
+    metac->set_key(metac, tieredname);
+    WT_ERR_NOTFOUND_OK(metac->search(metac), true);
+    if (ret == 0)
+        goto done;
+
+    /* Check colgroup metadata entry exists for uri. */
+    len = strlen("colgroup:") + strlen(format_name) + 1;
+    WT_ERR(__wt_calloc_def(r->session, len, &cgname));
+    WT_ERR(__wt_snprintf(cgname, len, "colgroup:%s", format_name));
+    metac->set_key(metac, cgname);
+    WT_ERR_NOTFOUND_OK(metac->search(metac), true);
+    if (ret == WT_NOTFOUND)
+        has_colgroup = false;
+
+    /* Check file metadata entry exists for uri. */
+    len = strlen("file:") + strlen(format_name) + strlen(".wt") + 1;
+    WT_ERR(__wt_calloc_def(r->session, len, &filename));
+    WT_ERR(__wt_snprintf(filename, len, "file:%s.wt", format_name));
+    metac->set_key(metac, filename);
+    WT_ERR_NOTFOUND_OK(metac->search(metac), true);
+    if (ret == WT_NOTFOUND)
+        has_file = false;
+
+    /* If either colgroup or file metadata entry doesn't exist mark the table for removal. */
+    if (!has_colgroup || !has_file) {
+        WT_ERR(__wt_realloc_def(
+          r->session, &r->remove_uris_allocate, r->nremove_uris + 1, &r->remove_uris));
+        WT_ERR(__wt_strdup(r->session, format_name, &r->remove_uris[r->nremove_uris]));
+        r->nremove_uris++;
+    }
+    ret = 0;
+done:
+err:
+    if (tieredname != NULL)
+        __wt_free(r->session, tieredname);
```

Comment:
`__wt_free` allows a NULL argument (for which it does nothing), so the `if` conditionals here are not needed.